docs: cosmetic cleanups (#11322)
snapcore/snapd: HACKING.md

tested and we use the [spread](https://github.com/snapcore/spread)
integration test framework for the integration/system level tests.

## Setting up

### Supported Ubuntu distributions

Ubuntu 18.04 LTS or later is recommended for `snapd` development.
Usually, the latest LTS would be the best choice.

>If you want to build or test on older versions of Ubuntu, additional steps
may be required when installing dependencies.

### Supported Go version

> [Go 1.13](https://pkg.go.dev/go@go1.13) or later is required to build `snapd`.

Author: Regarding the link issue, I am not sure I understand. This link brings you to the documentation of this version on the `pkg.go.dev`. The idea was to spare the reader (should they attempt searching for what Go 1.13 really is), since this link will bring them to the exact place. I thought it's useful especially since I mention another version below, and Go 1.13 is not the latest version of Go.

Regarding _the_ thing, people with Russian origins are known to be horrible with all this "article" thing :-). But jokes aside, I was told by one technical writer that in titles and presentation bullets it is acceptable to omit article of any sort, since it makes the statement nominal. But I can be wrong (@degville could you comment?). In any way, I don't see any problem with adding _the_. :-)

>If you need to build older versions of snapd, please have a look at the file
[debian/control](debian/control) to find out what dependencies were needed at the time
(including which version of the Go compiler).

> ### Setting up build environment

Author: These two lines were brought by diff by accident, since somewhere here in the previous version was a section with the similar heading, apparently.  

Regarding _your_ or _the_, I personally prefer _the_. First of all, _your_ form doesn't appear in any of the headings in this document. Second, it feels to me that the heading reflects a purpose not to declare a request. I mean when we give guidelines, it is acceptable to provide guidelines in a form like: "do this", "take your", etc. From my experience (as far as I understand it) this form is less preferable for headings.  

@mardy  and @MiguelPires, would it be okay with you if I added _the_, not _your_?

> ### Setting up build environment

Reviewer: I'm fine with either :-)

`snapd` is published and maintained as a [Go module](https://go.dev/doc/modules/managing-source). 
A module is a collection of [Go packages](https://go.dev/ref/spec#Packages) 
stored in a file tree with a [go.mod](go.mod) file at its root. 

> The [go.mod](go.mod) file defines the module’s _module path_, 

Author: Wow! Oops, it was a copy paste mistake - I took it from the Go documentation. :-(
Wow again, for the strength of your sight. You have the eyes of an eagle. :-)

I'll replace it.

which is also the import path used for the root directory, 
and its _dependency requirements_. 
All this allows maintaining `snapd` outside of the strict hierarchy of `GOPATH`, 
that was mandatory before the release of [Go 1.13](https://pkg.go.dev/go@go1.13)

> (see [build @Go1.10](https://pkg.go.dev/go/build@go1.10)).  

Author: I wasn't, and the Go documentation wasn't really straightforward on this topic. On contrary, in this area the documentation was very quite confusing for me.   

@MiguelPires, I think that this document is not for the elders. Though, I think not even all elders know all the details - simply because they don't need it. The fresh ones (who actually start from reading this document) do need this and also proper links to have more details that just one sentence in order to understand how this things work (in my opinion of course).

By the way, even Go documentation is written by Go pros, who not always can see what are the difficulties of the newbies.

Even with Go modules mode, `GOPATH` is still used by the Go tool implicitly, 
to store downloaded source code and compiled commands 
(see [Go docs](https://pkg.go.dev/cmd/go#hdr-GOPATH_and_Modules)).

> If `$GOPATH` is not set explicitly, in Unix-like systems it defaults to `$HOME/go`.

One can easily find the location of the `GOPATH` file-tree on any enabled machine
by running the following command:

```
go env GOPATH
```

#### Backward compatibility

Some tools we use still rely on `$GOPATH/bin` being listed in `$PATH` 
(see [Running spread with QEMU](#running-spread-with-qemu)).
Add `$GOPATH/bin` to your `PATH` in your shell login script 

> (e.g. `~/.bashrc` for bash):

Author: It was my lousy attempt to preserve as much of the sentences from the previous version. I see what you say - I'll fix it.

```sh

> export PATH=$PATH:$(go env GOPATH)/bin

Reviewer: Mmm... but where does `go env` get the value from? If it still uses `GOPATH` then using the environment variable directly will be much faster (we should try to keep the user's `.bashrc` execution as quick as possible).

Author: This is exactly the change that comes with Go1.13. `go` tool doesn't rely on `GOPATH` variable. It has a default value for this parameter, which can be overwritten with `GOPATH` environment variable. However, this method (i.e. environment variable), per my understanding of the Go documentation, is considered deprecated and is maintained primarily to support older projects that started before Go1.13 and haven't (or couldn't) migrate to the new mode of work (i.e. Modules).
By the way, it has no effect on performance whether you have this environment variable defined or not. Moreover, as I understood this variable might have negative effect (if explicitly defined) on certain modern tools that anticipate things in the default location (bad excuse - I agree, but this stuff happens).

Reviewer: Sorry, I was not very clear about the performance issue: I mean that the `go env` command takes time to execute, and if you put this in the `.bashrc` it will slow down the prompt. By a few milliseconds only, sure, but some time ago I had to go through my `.bashrc` and remove a few entries because it had grown to the point where the delay was not just noticeable, but annoying.

Anyway, this is not a big issue.

```

If your `GOPATH` contains more than one path, you need to add `/bin`
suffix to each path:

```sh
export PATH=$PATH:$(go env GOPATH | sed -e "s|:|/bin:|g")/bin
```

> #### Setting custom `GOPATH` (_Optional_)

Reviewer: I would actually vote for keeping this the suggested way, since it makes everything easier. I think we could make this document a bit slimmer if we just recommended this way (advanced users will find their own path anyway).

Author: Look, here I tend to disagree, once again since the `GOPATH` thing is considered a deprecated mod of operation. 
Moreover, if there are older projects that some of our contributors might use, who need this variable configured - we make their life more complicated without an actual need in it. 

Moreover, if any I would have recommended removing this section, if we are willing to make this document slimmer. It will also simplify the process of installation. If I don't need to configure something, why configuring it.

Reviewer: I think we should write the minimal set of instructions to get the contributor's environment working. I agree that if `GOPATH` is no longer necessary, then we should not mention it.

Though it's no longer required, one still may set a custom location for `GOPATH`.

```sh
export GOPATH=${HOME}/work
```

This change can be done permanently either using Linux conventional way

> (see the `$GOPATH/bin` [example](#backward-compatibility)) or using the Go tool:

Reviewer 2: I'm not opposed to suggesting we install spread in the conventional bin folders, but also maybe continuing my other line of thinking perhaps we should punt on the issue entirely and just say that folks need to install spread and link to upstream spread's installation instructions which may or may not include using GOPATH. 

Also to be clear, I am more opposed to explaining how GOPATH works and how to set it up in depth, I'm not opposed to a single line note that says if you install snapd or spread into $GOPATH/bin then you need to add that to your $PATH somehow, that seems reasonable to me.

```sh
go env -w GOPATH=${HOME}/work
```

> ### Getting snapd sources

Reviewer: I liked it better before :-)

Author: Sure, I'll revert it.

    cd snapd

This will allow you to build and test `snapd`. If you wish to contribute to
the `snapd` project, please see the [Contributing section](#contributing).

> For more details about source-code structure of `snapd` please read about
[Managing module source](https://go.dev/doc/modules/managing-source) in Go.

### Installing the build dependencies

Build dependencies can automatically be resolved using `build-dep` on Ubuntu:

    cd ~/snapd
    sudo apt-get build-dep .

Package build dependencies for other distributions can be found under the
[./packaging/](./packaging/) directory.

Source dependencies are automatically retrieved at build time.

> Sometimes, it might be useful to pull them without the very building:

Reviewer 3: ```suggestion
Sometimes, it might be useful to pull them without building:
```

```
cd ~/snapd
go get ./...
```

> 

Author: Sure. Thank you.

can either use `snap revert snapd`, or you can refresh directly with 
`snap refresh snapd --stable --amend`.

> It is also sometimes useful to use snapcraft to build the snapd snap for
other architectures using the `remote-build` feature, however there is currently a
bug in snapcraft around using the `4.x` track and using `remote-build`, where the

> Launchpad job created for the remote-build will attempt to use the `latest` track instead

Author: So, just to make sure I understand you correctly, @anonymouse64. 

You suggestion is to remove the text starting from `However, this fix does...` (until the end of the paragraph). 
since this explicit removal of these files is not longer needed. Do I understand correctly?

If yes, then I also think that we need to remove the following command line too, as after the fix it's done automatically:
```
rm -r tests/main/validate-container-failures/
```
Am I right?

> Launchpad job created for the remote-build will attempt to use the `latest` track instead

Reviewer 2: Yes that is all correct

    sudo apt-get build-dep .

Package build dependencies for other distributions can be found under the
[./packaging/](./packaging/) directory.

Go module dependencies are automatically resolved at build time.
One can do it directly, without building, using the following commands:

```
cd ~/snapd
go get ./...
```

## Building

> ### Building the snap with snapcraft <!-- TODO: (A) Is this section outdated??? -->

Author: Thank you. 

@anonymouse64, I have the question then. In this chapter I've seen that Ubuntu 16.04 is used. Didn't we move to 18.04?
And regarding the workaround that you described in this chapter for the bug in Snapcraft, is the fix for the bug is still not merged to the stable? Do we still need the workaround?

> ### Building the snap with snapcraft <!-- TODO: (A) Is this section outdated??? -->

Reviewer 2: As I mention in my other comment, in order to build the snap natively with `--destructive-mode` you NEED to use Ubuntu 16.04 - in other words for this specific action (building the snap with specifically `--destructive-mode`) no we have not moved yet

The easiest (though not the most efficient) way to test changes to snapd is to
build the snapd snap using _snapcraft_ and then install that snapd snap. The
[snapcraft.yaml](./build-aux/snapcraft.yaml) for the snapd snap is located at 
[./build-aux/](./build-aux/), and
can be built using snapcraft either in a LXD container or a multipass VM (or

> natively with `--destructive-mode` on a Ubuntu 18.04 host).

Author: @anonymouse64, the same as above, haven't we moved to 18.04?

> natively with `--destructive-mode` on a Ubuntu 18.04 host).

Reviewer 2: no

snap refresh snapcraft --channel=4.x/stable
```

#### Splicing the snapd snap into the core snap

Sometimes while developing you may need to build a version of the _core_ snap
with a custom snapd version. 

> The `snapcraft.yaml` for the [core snap](https://github.com/snapcore/core20/) 

Reviewer 2: this link is wrong it should be to https://github.com/snapcore/core (and to be clear no, it is not updated to core20, that is a different snap, this is very specifically for the core snap which is special and distinct from the core20 snap)

Reviewer 3: I don't think this was fixed

Reviewer 3: I pushed a PR change the link https://github.com/snapcore/snapd/pull/11368

#### Splicing the snapd snap into the core snap

Sometimes while developing you may need to build a version of the _core_ snap
with a custom snapd version. 

> The [snapcraft.yaml](https://github.com/snapcore/core20/blob/master/snapcraft.yaml) 

Author: OMG. I am so sorry for this one. I'll fix it.

  Tag_CPU_arch: v7
  Tag_FP_arch: VFPv3-D16
```

## Contributing

Contributions are always welcome!

Please make sure that you sign the Canonical contributor agreement [here](
http://www.ubuntu.com/legal/contributors).

Complete requirements can be found in [CONTRIBUTING.md](./CONTRIBUTING.md).

### Guidelines for contributors

There are few guidelines that contributors should follow:

> - Commit messages should be well structured,  

Author: @MiguelPires, I am not sure I agree. Once again it's not a rule here, but just a recommendation. It's very simple, please look at it.

> - Commit messages should be well structured,  

Reviewer 3: > @MiguelPires, I am not sure I agree. Once again it's not a rule here, but just a recommendation. It's very simple, please look at it.

You don't agree with changing ',' to '.' ? Also, what does "it's very simple, please look at it" mean? What did I not look at?

> - Commit messages should be well structured,  

Author: @MiguelPires, I assumed that you suggest to put here "." since you suggest to omit the reference to the the [Conventional Commits](https://www.conventionalcommits.org/). Therefore, I asked you to look at it, I mean at the specification.

> - Commit messages should be well structured,  

Reviewer 3: No, it's just changing the comma for the period since every other list item uses a period

> - Commit messages should be well structured,  

Author: As I expected, I simply didn't understand you right. :flushed: 

Buy the way, I would love to hear you opinion on this specification?:smiley:

> (_Consider complying with the [Conventional Commits](https://www.conventionalcommits.org/) specification_)

Author: @MiguelPires, we actually had an example here of how commit should like, which in lack of other definition serves as a convention. When I tried to format it a little, I was suggested here in review to point to some other sources that are updated and can be referred instead. I found something that is widely used and referred by others, which is very simplistic with no severe restrictions of any sort. It doesn't conflict with format of messages in snapd.

But the most important thing, as you might see, it is not a rule, but a thing that one may consider. It might help those who feel uncomfortable with guessing how should the message look like. If you want you can follow this conventions, if you don't then it's up to you. However, if you are confused and need some good example, this is definitely a good concise and very simple one. Wouldn't you agree?

> (_Consider complying with the [Conventional Commits](https://www.conventionalcommits.org/) specification_)

Reviewer 3: To be clear, I'm not saying people shouldn't use conventional commits for snapd. I just think it's strange to recommend something that (AFAIK) we don't do. Usually, we use `pkg/sub: did a thing" or just "did a thing", which people can see in snapd's git history/open PRs. Saying "Consider doing this. We don't do it but you can" seems confusing to me. But that's just my opinion, I may be wrong. If you think having this is more welcoming to newcomers to the project, then feel free to leave it :+1:

> (_Consider complying with the [Conventional Commits](https://www.conventionalcommits.org/) specification_)

Reviewer 2: I agree we should not recommend things we don't actually use as a team

> (_Consider complying with the [Conventional Commits](https://www.conventionalcommits.org/) specification_)

Author: Guys, I understand your point. At the same time I believe we need at least some guidelines. Perhaps, it shouldn't be part of this PR and I need to bring this to Michael. I am talking from the impression that I got through my experience in these topics. If might sound weird to you, but some people that have good experience still don't know that there is a difference between description and the title inside the git message and you are supposed to place an empty line between these to. There are also some "unspoken" conventions that are not known to those who come from other VCSs (yes, there are still companies that using VCSs other than git :smiley: ).

> (_Consider complying with the [Conventional Commits](https://www.conventionalcommits.org/) specification_)

Reviewer 2: Sure, in this case remove it from the PR and bring the suggestion up to mvo and then we can discuss using those guidelines (or any other guidelines) specifically in a separate PR, that will unblock you from landing the rest of the changes in this PR

- Commit emails should not have non-ASCII characters.  
- PR title should start with a prefix denoting the 

> `<feature>`/`<package>`/`<module>`/`<file modified>` following by a semicolon.    

Author: @MiguelPires what about the change that spread over several packages, I mean at feature level? What about things that are not related to any package at all (like this PR)? Shouldn't we list at least both "package" and "feature".

> `<feature>`/`<package>`/`<module>`/`<file modified>` following by a semicolon.    

Author: For someone like me (over pedantic), it makes it more clear to see these options. So, if you don't mind, I do prefer to leave it here. :pray:

  Tag_CPU_arch: v7
  Tag_FP_arch: VFPv3-D16
```

## Contributing

Contributions are always welcome!

Please make sure that you sign the Canonical contributor agreement [here](
http://www.ubuntu.com/legal/contributors).

Complete requirements can be found in [CONTRIBUTING.md](./CONTRIBUTING.md).

### Guidelines for contributors

There are few guidelines that contributors should follow:

> - Commit messages should not have non-ASCII characters.  

Author: Got it. I'll fix it.

> (_otherwise, the PR will fail_)

Author: Got it.

- Commit messages should be well structured,  
(_Consider complying with the [Conventional Commits](https://www.conventionalcommits.org/) specification_)

> - PR title should start with prefix denoting the feature/module following by semicolon.  

Author: Got it

(_otherwise, the PR will fail_)
- Try to open more smaller PR's, rather than one large PR.
- Try not to mix controversial and trivial changes together.  
  (_Proposing trivial changes separately makes landing-them easier and 
  makes reviewing the controversial changes simpler_)
- Try not to force push to PR's after they have gotten reviews.
- Try to write tests to cover the contributed changes.

>If you need any help with any of these guidelines, please reach out to the team.

### Creating Pull Requests

Contributions are submitted through a 
[Pull Request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork) 
created from a 
[fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo#forking-a-repository) 

> of the [snapd repository](https://github.com/are-scenic/snapd) (under your GitHub account).

Author: Got it

>This complete process is outlined in details in the [GitHub documentation](

> https://docs.github.com/en/github/collaborating-with-pull-requests). 

Author: Yes, it does. I constantly check the version in a browser and in vscode preview (one of the strengths of this IDE:-) )

directory and:

```
go test -check.f $testname
```

If a test hangs, you can enable verbose mode:

```
go test -v -check.vv
```

Or, try just `-check.v` for a less verbose output.

> Some unit tests are known to fail on locales other than `C.UTF-8`. 
If you have unit tests failing, try setting `LANG=C.UTF-8` when running 
`go test`. See [issue #1960131](https://bugs.launchpad.net/snapd/+bug/1960131) for more details.

There is more to read about the testing framework on the [website](https://labix.org/gocheck)

### Running integration tests

#### Downloading spread framework

To run the integration tests locally via QEMU, you need the latest version of
the [spread](https://github.com/snapcore/spread) framework. 
You can get spread, QEMU, and the build tools to build QEMU images with:

    $ sudo apt update && sudo apt install -y qemu-kvm autopkgtest
    $ curl https://storage.googleapis.com/snapd-spread-tests/spread/spread-amd64.tar.gz | tar -xz -C <target-directory>

> `<target-directory>` can be any directory that is listed in `$PATH`, 
as it is assumed further in the guidelines of this document. 
You may consider creating a dedicated directory and adding it to `$PATH`, 

> or you may chose using one of the conventional Linux directories (e.g. `/usr/local/bin`)

Reviewer 3: ```suggestion
or you may choose to use one of the conventional Linux directories (e.g. `/usr/local/bin`)
```

You can run an individual test for a sub-package by changing into that 
directory and:

```
go test -check.f $testname
```

If a test hangs, you can enable verbose mode:

```
go test -v -check.vv
```

Or, try just `-check.v` for a less verbose output.

> > `snapd` before printing any text adjusts it to a dialect that
> corresponds to the locale of the hosting system. For instance, 
> texts for `en_US` and for `en_GB` not always look alike.
> Unfortunately, some tests don't take it into account (for the details 
> see [issue #1960131](https://bugs.launchpad.net/snapd/+bug/1960131)).
> A temporary workaround is to specify `LANG=C.UTF-8 ` in the command 
> line right before `go test`, if you experience similar issues.

Reviewer 3: Just a suggestion, slightly more concise
```suggestion
> Before printing text, `snapd` translates it according to the 
system's locale. Currently, some unit tests don't take this into 
account (see [issue #1960131](https://bugs.launchpad.net/snapd/+bug/1960131)), which might
cause tests to fail on locales other than US English.
A temporary workaround is to set `LANG=C.UTF-8 ` when running `go test`.
```

Author: ```suggestion
> Before printing text, `snapd` adjusts it according to the 
system's locale. Currently, some unit tests don't take this into 
account (see [issue #1960131](https://bugs.launchpad.net/snapd/+bug/1960131)), which might
cause tests to fail on locales other than US English.
A temporary workaround is to set `LANG=C.UTF-8 ` when running `go test`.
```

@MiguelPires, I don't think translate is the right word here, as it makes really small adjustments, far from real translation. Therefore, I recommend to use verb "adjust" instead of "translates". 
Would it be okay with you?

Reviewer 3: That's only the case if your locale is something like en_GB, in which case the adjustments are small (e.g., "color" -> "colour"). If your locale is an entirely different language, then it will translate it (depending on available translations).

You can run an individual test for a sub-package by changing into that 
directory and:

```
go test -check.f $testname
```

If a test hangs, you can enable verbose mode:

```
go test -v -check.vv
```

Or, try just `-check.v` for less verbose output.

> > Certain locales imply adjustments in texts printed by snapd 

Author: @anonymouse64, I am not sure I fully understand what you mean and how LP is related to it. I observed this failure on my laptop that has such locale.  
There is a number of tests that have hard-coded string, whereas snapd processes it's responses to adjust to the locale and then prints it out. 
By the way, following your request I've identifies the list of these tests (please see the [top comment](https://github.com/snapcore/snapd/pull/11322#issue-835257404) in this PR).

> > Certain locales imply adjustments in texts printed by snapd 

Reviewer 2: I mean please create a new bug on launchpad at bugs.launchpad.net/snapd with the list of tests that fail with your locale, and then include a link to that bug right here

> > Certain locales imply adjustments in texts printed by snapd 

Author: @anonymouse64, I see. I think that there is one point that I failed to clarify. I think it's not an issue of LP, but of these specific tests. This tests assume a specific text to be printed on STDOUT, which they capture to verify it. The tests use hard-coded strings to compare with captured output from STDOUT. The tests don't take into account that each message before being printed to STDOUT is preprocessed to adjust to locale (please look at the sub-directory `./po/`. I mean this thing is controlled by us, per my understanding.

> > Certain locales imply adjustments in texts printed by snapd 

Reviewer 2: @are-scenic I mean that these tests were not written correctly - they should pass and work if they are run with your locale or my locale or any other locale out there. These tests should _not_ depend on the user's locale, they should be verifying that we use the right messages with a specific, tested locale. I have not reviewed this specific list of tests and it's unlikely I will any time soon, but I would be surprised if these tests are in fact testing that the locale/translations are correct / working. I don't think those tests should be testing that, it's not the appropriate level.

All that is to say IMHO those tests are broken right now, and the thing we should do is fix those tests so they are always passing regardless of what locale the user is using. 

So what I am telling you to do is to file a new launchpad bug (where we track bugs for snapd), which indicates that these tests fail when the user's locale is not the one the tests were originally written with (I have already forgotten which locale that is from this extended conversation). Then after filing such a bug, leave a comment here in the HACKING.md which explains that if some of the tests fail for a user and their locale is not the expected locale, that they are running into the bug which you filed and that to work around that bug, they should define the environment variable to set the right locale so the tests pass

> > Certain locales imply adjustments in texts printed by snapd 

Author: @anonymouse64, I am so sorry. I feel really stupid right now. I simply didn't understand your all along. :flushed:  
I thought you are implying that there is a bug in LP itself, which I fail to understand. But what you meant was actually to file a new bug for `snapd` and that we are using LP as a register of bugs we find in `snapd`.

By the way, I believe we shall mention this in this doc too. Those who contribute shall know how and where to file a bug if they suspect they've found one.

```

Make changes to the repository and commit the changes to your
local working branch.

```sh
git commit -sa -m "<feature/subject-prefix>: <brief-description-of-the-change>

<more-elaborate-description-with-reasoning-about-the-changed>"
``` 

> >If it's justified, one might consider several commits before pushing 

Author: I totally understand. I need your help to understand one thing then. Do we have certain guidelines that are specific to us?

For instance, PR caption has to have format: "module/feature: caption", and I actually failed on my PR because of it. Are there any other requirements like these? 
It's not mandatory for GitHub or git, but if we verify as part of PR, we need to state it clear here, or in some other dedicated document.

> >If it's justified, one might consider several commits before pushing 

Reviewer 2: yes we should also mention the PR title checker and probably also that commit messages should not have non-ASCII characters as their email addresses like the recent checker that we added for this, other than that no I don't think we have other guidelines to document

their work to the remote repository. 
One should strive to have only one commit per topic, and to avoid pushing 

> many topics in the same attempt.  

Reviewer: s/attempt/feature branch/

Author: @mardy, I am sorry, but I don't understand the comment. :-[] Could you please explain? :-)

Reviewer 3: It means replace "attempt" with "feature branch"

Author: Is it VIM syntax?

Reviewer: I think it comes from the `sed` command, but yes, it's also present in ViM :-)

Push the changes to your forked `snapd` repository.

```
git push fork <local-branchname>[:<remote-branchname>]
```

>Note that if the `<remote-branchname>` is not specified, a new remote branch 

> is created implicitly, by the name of the `<local-branchname>`.

Reviewer 2: I also agree with Miguel, I don't think we needed to have all this originally, IMHO all the explanation of how to make a git commit is not appropriate here, again I'm fine to leave links to some upstream documentation / tutorial but it seems too much detail for us to maintain and will just inevitably lead to debate like this about how detailed we want to be about documenting how to use git

> is created implicitly, by the name of the `<local-branchname>`.

Author: @MiguelPires  and @anonymouse64, are there any specific things that we assume as part of a PR, perhaps even verify for it? If so, do we have a documentation for it to refer to?
Are there any other good practices that we assume our contributors shall follow? Do we have links to these somewhere?

Create the Pull Request for your branch on GitHub.

This complete process is outlined in the GitHub documentation [here](
https://docs.github.com/en/github/collaborating-with-pull-requests).

## Testing

> ### Running unit-tests

Reviewer: I would move this after the next paragraph.

Author: What do you mean? Moving unit-tests after integration tests?

Author: Sure, I get it. You're right. The first paragraph of the section should address the whole set of testing, not just unit-testing. My bad. Thank you, for pointing it out.

to the official list of [Ubuntu releases](https://wiki.ubuntu.com/Releases). 

> `<release-short-name>` is the first word in the release's full name, 
e.g. for "Bionic Beaver" it is `bionic`.

To build an Ubuntu 14.04 (Trusty Tahr) based VM, use:

    $ autopkgtest-buildvm-ubuntu-cloud -r trusty --post-command='sudo apt-get install -y --install-recommends linux-generic-lts-xenial && update-grub'
    $ mv autopkgtest-trusty-amd64.img ubuntu-14.04-64.img

> This is because we need at least 4.4+ kernel for snapd to run on Ubuntu 14.04 
LTS, which is available through the `linux-generic-lts-xenial` package.

If you are running Ubuntu 16.04 LTS, use 
`adt-buildvm-ubuntu-cloud` instead of `autopkgtest-buildvm-ubuntu-cloud` (the
latter replaced the former in 18.04):

    $ adt-buildvm-ubuntu-cloud -r xenial
    $ mv adt-<release-name>-amd64-cloud.img ubuntu-<release-version>-64.img

#### Downloading spread VM images

Alternatively, instead of building the QEMU images manually, you can download
pre-built and somewhat maintained images from 
[spread.zygoon.pl](https://spread.zygoon.pl/). The images will need to be extracted 
with `gunzip` and placed into `~/.spread/qemu` as above.

> > An image for Ubuntu Core 20 that is pre-built for KVM can be download from 

Reviewer 3: ```suggestion
> An image for Ubuntu Core 20 that is pre-built for KVM can be downloaded from 
```

#### Downloading spread VM images

Alternatively, instead of building the QEMU images manually, you can download
pre-built and somewhat maintained images from 
[spread.zygoon.pl](https://spread.zygoon.pl/). The images will need to be extracted 
with `gunzip` and placed into `~/.spread/qemu` as above.

#### Running spread with QEMU

Finally, you can run the spread tests for Ubuntu 18.04 LTS 64-bit with:

    $ spread -v qemu:ubuntu-18.04-64

>To run for a different system, replace `ubuntu-18.04-64` with a different system
name, which should be a basename of the [built](#building-spread-vm-images) or 
[downloaded](#downloading-spread-vm-images) Ubuntu image file.

For quick reuse you can use:

    $ spread -reuse qemu:ubuntu-18.04-64

It will print how to reuse the systems. Make sure to use
`export REUSE_PROJECT=1` in your environment too.

#### Running UC20 spread with QEMU

To install a VM of Ubuntu Core 20 refer to the [official tutorial](https://ubuntu.com/download/kvm). 
It provides a ready UC20 image to download and has the instructions for prerequisites, 
like installing OVMF package (to meet the requirement for UEFI).  

> > If needed, you may also [create your own image](https://snapcraft.io/tutorials/create-your-own-core-20-image) of Ubuntu Core 20.  

Author: @anonymouse64, here I really need your help to understand this one. Are you talking about the `> If needed, you may...` part or the entire chapter?  
I mean the chapter exists in the original [HACKING.md](https://github.com/snapcore/snapd/blob/master/HACKING.md#running-uc20-spread-with-qemu).

> > If needed, you may also [create your own image](https://snapcraft.io/tutorials/create-your-own-core-20-image) of Ubuntu Core 20.  

Reviewer 2: I think we still ought to mention somewhere/somehow that folks need to install ovmf package in order to run the spread tests for UC20 locally, but you're correct in that actually there is no difference after that point running spread tests against UC20 vs all other systems

> > If needed, you may also [create your own image](https://snapcraft.io/tutorials/create-your-own-core-20-image) of Ubuntu Core 20.  

Author: @anonymouse64, what about the env-variable `SPREAD_QEMU_BIOS`, shouldn't it be set to `/usr/share/OVMF/OVMF_CODE.fd` anymore?

> > If needed, you may also [create your own image](https://snapcraft.io/tutorials/create-your-own-core-20-image) of Ubuntu Core 20.  

Reviewer 2: AFAICT that is not needed or used at all anymore

Now you are ready to run:

>     $ SPREAD_QEMU_BIOS=/usr/share/OVMF/OVMF_CODE.fd spread -v qemu:ubuntu-core-20-64

Author: @anonymouse64, if I read you and the spread code you mention correctly, we have no issues with running spread on UC20, and this chapter can be removed.   

Do I understand correctly?

>     $ SPREAD_QEMU_BIOS=/usr/share/OVMF/OVMF_CODE.fd spread -v qemu:ubuntu-core-20-64

Reviewer 2: as mentioned in the other comment, the only thing we need to mention here is just that one has to install ovmf to run spread tests against UC20

You can run an individual test for a sub-package by changing into that 
directory and:

```

> LANG=C.UTF-8 go test -check.f $testname

Reviewer: mmm... is `.UTF-8` needed?

Author: @mardy , as a person who lives in GB and has a system locale that is not US, I assure you that this one is critical. :-) When it fails for someone (especially fresh) who has a different locale, finding the reason is even far beyond from the definition of "painful", especially if you have no debugger (which I hope to resolve soon too:-)).  

By the way, I've added a comment explaining the reasoning.

Reviewer 2: @are-scenic it would be great if you can explain which tests fail and how (not in this PR), we should probably fix those tests so that they work on all locales (i.e. the test itself sets a specific expected locale when executing) without having to specify this

Author: Sure I can. I will identify them and provide you with the list.

```

If a test hangs, you can enable verbose mode:

```
LANG=C.UTF-8 go test -v -check.vv
```

(or -check.v for less verbose output).

> >Note `LANG=C.UTF-8` is specified in the examples above only to

Author: @MiguelPires, I'll try to reflect your proposal. If I understand you correctly, you suggest to omit the explicit locale specification from the example, but at the same time you think it's better to extend the clarification that some tests might fail as they assume specific local, which not always might be the case. Do I understand correctly?

If I do understand, will it be okay with you if I add into clarification an advice to add `LANG=C...` before the commend if someone experiences this phenomenon?

> >Note `LANG=C.UTF-8` is specified in the examples above only to

Reviewer 2: agreed with Miguel

> ensure locale compatibility for some unit-tets. Some locales 

Author: Thank you.

To debug interaction with the snap store, you can set `SNAP_DEBUG_HTTP`.
It is a bitfield: dump requests: 1, dump responses: 2, dump bodies: 4.

> In case you get some security profiles errors, when trying to install or refresh a snap, 
maybe you need to replace system installed snap-seccomp with the one aligned to the snapd that 
you are testing. To do this, simply backup `/usr/lib/snapd/snap-seccomp` and overwrite it with 

> the testing one. Don't forget to roll back to the original, after you finish the testing.

Reviewer 3: ```suggestion
the testing one. Don't forget to roll back to the original, after you finish testing.
```

- _Google_: `spread google-nested:ubuntu-20.04-64:tests/nested/core20/tpm`  

The nested system in all the cases is selected based on the host system. The following lines 
show the relation between host and nested `systemd` (same applies to the classic nested tests):

- ubuntu-16.04-64 => ubuntu-core-16-64
- ubuntu-18.04-64 => ubuntu-core-18-64
- ubuntu-20.04-64 => ubuntu-core-20-64

The tools used for creating and hosting the nested VMs are:

- _ubuntu-image snap_ is used to build the images
- _QEMU_ is used for the virtualization (with [_KVM_](https://www.linux-kvm.org/page/Main_Page) acceleration)

Nested test suite is composed by the following 4 suites:

- _classic_: the nested suite contains an image of a classic system downloaded from cloud-images.ubuntu.com 
- _core_: it tests a core nested system, and the images are generated with _ubuntu-image snap_
- _core20_: this is similar to the _core_ suite, but these tests are focused on UC20
- _manual_: tests on this suite create a non generic image with specific conditions

The nested suites use some environment variables to configure the suite 
and the tests inside it. The most important ones are described bellow:

> - `NESTED_WORK_DIR`: it is path to the directory where all the nested assets and images are stored

Reviewer 3: ```suggestion
- `NESTED_WORK_DIR`: path to the directory where all the nested assets and images are stored
```

- `NESTED_TYPE`: use core for Ubuntu Core nested systems or classic instead.
- `NESTED_CORE_CHANNEL`: the images are created using _ubuntu-image snap_, use it to define the default branch
- `NESTED_CORE_REFRESH_CHANNEL`: the images can be refreshed to a specific channel; use it to specify the channel
- `NESTED_USE_CLOUD_INIT`: use cloud init to make initial system configuration instead of user assertion
- `NESTED_ENABLE_KVM`: enable KVM in the QEMU command line
- `NESTED_ENABLE_TPM`: re-boot in the nested VM in case it is supported (just supported on UC20)

> - `NESTED_ENABLE_SECURE_BOOT`: enable secure boot in the nested VM in a case it is supported (supported just on UC20)

Reviewer 3: ```suggestion
- `NESTED_ENABLE_SECURE_BOOT`: enable secure boot in the nested VM in case it is supported (supported just on UC20)
```

    $ mkdir -p ~/.spread/qemu
    $ cd ~/.spread/qemu

Assuming you are building on Ubuntu 18.04 LTS ([Bionic Beaver](https://releases.ubuntu.com/18.04/)) 
(or a later Ubuntu), run the following to build a 64-bit Ubuntu 16.04 LTS 
(or later) with the correct `<release-name>` and `<release-version>`:

    $ autopkgtest-buildvm-ubuntu-cloud -r <release-name>
    $ mv autopkgtest-<release-name>-amd64.img ubuntu-<release-version>-64.img

Use the following table as an example reference:

| Release   | `<release-name>`  | `<release-version>`   |
| -------   | --------------    |-------------------    |
| [Xenial Xerus](https://releases.ubuntu.com/16.04/)  | xenial  | 16.04 |
| [Bionic Beaver](https://releases.ubuntu.com/18.04/) | beaver  | 18.04 |

> | [Focal Fossa](https://releases.ubuntu.com/20.04/)   | focal   | 20.04 |

Reviewer 2: this to me again seems like an example of documenting something we don't need to maintain here, I would rather us link to an upstream page which has this table - for example this does not mention impish or jammy here which we now have in our spread test environment and folks might want to use, so having to update this table all the time seems worse than just not having it here

Author: @anonymouse64, even better. I love it. Could you please refer me to the right location?

Reviewer 2: an extremely cursory google result reveals https://wiki.ubuntu.com/Releases as a perfectly reasonable table of such values

#### Running spread with QEMU

Finally, you can run the spread tests for Ubuntu 18.04 LTS 64-bit with:

    $ spread -v qemu:ubuntu-18.04-64

>To run for a different system, replace `ubuntu-18.04-64` with a different system
name, which should be a basename of the [built](#building-spread-vm-images) or 
[downloaded](#downloading-spread-vm-images) Ubuntu image file.

For quick reuse you can use:

    $ spread -reuse qemu:ubuntu-18.04-64

It will print how to reuse the systems. Make sure to use
`export REUSE_PROJECT=1` in your environment too.

#### Running UC20 spread with QEMU

> <!-- TODO: (B) Is it still working this way? -->

Reviewer 2: it would be great if you could update this section, indeed some things have landed upstream so it is not needed to build from the PR mentioned anymore

Author: Great. Thank you. @anonymouse64, from whom can I get the details to understand what is the updated way of doing it?

Reviewer 2: It looks like the updated way to do this is actually just to install ovmf and using spread from upstream AFAICT, so I would recommend just trying to install spread and then run some spread tests locally and see if it works

Author: Great. Thank you. I'll try this one.

using QEMU. For one, upstream spread currently does not support specifying what
kind of BIOS to use with the VM, so you have to build spread from this PR:
https://github.com/snapcore/spread/pull/95, and then use the environment 
variable `SPREAD_QEMU_BIOS` to specify an UEFI BIOS to use with the VM, for 
example the one from the OVMF package.  

>To get OVMF on Ubuntu, you can just install the `ovmf` package via `apt`.  

After installing OVMF, you can then run spread like so:

    $ SPREAD_QEMU_BIOS=/usr/share/OVMF/OVMF_CODE.fd spread -v qemu:ubuntu-core-20-64

This will enable testing UC20 with the spread, albeit without secure boot 
support. None of the native UC20 tests currently require secure boot however, 
all tests around secure boot are nested, see the section below about 
[running the nested tests](#running-nested-tests).

Also, due to the in-flux state of spread support for booting UEFI VM's like 
this, you can test `ubuntu-core-20-64` only by themselves and not with any other
system concurrently since the environment variable is global for all systems in
the spread run. This will be fixed in a future release of spread.

> ### Testing `snapd` daemon

Reviewer 3: This looks off to me, it should be "Testing the snapd daemon" or "Testing snapd" (the original version). IMO I'd go with the original

Author: The thing with _the_ is my weak point. :-)

I still have a question. From what I understood this section describes the tests that cover only `snapd` daemon. The entire package is named `snapd`. Other tests here also test `snapd`. 

Doesn't it require clarification?

Author: I prefer "Testing the snapd daemon", since these tests cover only daemon, whereas all tests referred by this document test `snapd`. I'll check this "_the_" thing with Graham.

Please run `(cd cmd; make fmt)` before sending your patches for the "C" part of
the source code.

<!-- !TODO: Few things to clean up in the future:

[] Add a section that describes functional labels in GitHub that we use to influence the verification flow of the PR

> [] Add some description about where (i.e. launchpad) to file a bug if we encounter one

Author: :flushed:, sorry my bad, 'completely forgot to check with it.

> [] Remove reference to https://bugs.launchpad.net/snapd/+bug/1960131 once it gets fixed

Author: This part is dropped when the markdown syntax is rendered. So, it's not seen. I thought that having some embedded centralized place will bring more spotlight to editor and keep it hidden from the reader.

@anonymouse64, if these comments are hidden and can serve as a single point of tracking the backlog of this document, would you consider leaving it here (minus unrelated launchpad bullet of course)?

> [] Remove reference to https://bugs.launchpad.net/snapd/+bug/1960131 once it gets fixed

Reviewer 2: Given that this doesn't render for folks reading the rendeded HACKING.md, I don't have a strong opinion, if you truly want to include it, that's fine, I just think it is redundant

To debug interaction with the snap store, you can set `SNAP_DEBUG_HTTP`.
It is a bitfield: dump requests: 1, dump responses: 2, dump bodies: 4.

> Note that in case you get some security profiles errors, when trying to install or refresh a snap, 
maybe you need to replace system installed snap-seccomp with the one aligned to the snapd that 
you are testing. To do this, simply backup `/usr/lib/snapd/snap-seccomp` and overwrite it with 

> the testing one. Don't forget to rollback to the original when finish testing

Reviewer 3: ```suggestion
the testing one. Don't forget to rollback to the original when you finish testing
```

Author: Thank you.
